model/minsky.cc

```diff
@@ -288,78 +288,87 @@ namespace minsky
-                         } 
-                     return false;
-                   });
-	}                              
+    if (!history.empty() || !canvas.model.get()->empty()) {     
+      bool alreadyDefinedMessageDisplayed=false;
+      
+      // convert stock variables that aren't defined to flow variables, and other fix up multiply defined vars
+      g->recursiveDo(&GroupItems::items,
+                     [&](Items&, Items::iterator i) {
+                       if (auto v=(*i)->variableCast())
+                         if (v->defined() || v->isStock())
+                           {
+                             // if defined, check no other defining variable exists
+                             auto alreadyDefined = canvas.model->findAny
+                               (&GroupItems::items,
+                                [&v](const ItemPtr& j)
+                                {return j.get()!=v && j->variableCast() &&  j->variableCast()->defined();});
+                             if (v->isStock())
+                               {
+                                 if (v->defined() && alreadyDefined && !alreadyDefinedMessageDisplayed)
+                                   {
+                                     message("Integral/Stock variable "+v->name()+" already defined.");
+                                     alreadyDefinedMessageDisplayed=true;
+                                   }
+                                 else if (!v->defined() && !alreadyDefined)
+                                   {
+                                     // need to do this var explicitly, as not currently part of model structure
+                                     if (auto vp=VariablePtr(*i))
+                                       {
+                                         vp.retype(VariableType::flow);
+                                         *i=vp;
+                                         convertVarType(vp->valueId(), VariableType::flow);
+                                       }
+                                   }
+                               }
+                             else if (alreadyDefined)
+                               {
+                                 // delete defining wire from this
+                                 assert(v->ports.size()>1 && !v->ports[1]->wires().empty());
+                                 g->removeWire(*v->ports[1]->wires()[0]);
+                               }
+                           }
+                       return false;
+                     });
+    }                              
 
     canvas.model->addGroup(g); // needed to ensure wires are correctly handled
     auto copyOfItems=g->items;
-
-    for (auto& i: copyOfItems)
-      {		
-         canvas.model->addItem(i);			  
-         canvas.selection.ensureItemInserted(i);
-         assert(!i->ioVar());
-      }
-
+    auto copyOfGroups=g->groups;
+    
+    g->resizeOnContents(); // used in insertGroupFromfile(), so why not here?
```

Comment:
Because this just changes the physical layout of the group, but that aspect is not used later on